README.md
Featherlight - ultra slim jQuery lightbox [![Build Status](https://travis-ci.org/noelboss/featherlight.svg?branch=master)](https://travis-ci.org/noelboss/featherlight)
------------------------

**Featherlight is a very lightweight jQuery lightbox plugin. For more information and demos, visit the [official website](http://noelboss.github.io/featherlight/).**

* Simple yet flexible
* Image, Ajax, iFrame and custom content support
* [Gallery Extension](https://github.com/noelboss/featherlight/#featherlight-gallery)
* Minimal CSS
* Name-spaced CSS and JavaScript
* Responsive
* Customizable via configuration object


## [» Download Current Release 0.4.4](https://github.com/noelboss/featherlight/archive/0.4.4.zip)

Here you'll find a [list of all the changes](https://github.com/noelboss/featherlight/blob/master/CHANGELOG.md) and you can also download [old releases](https://github.com/noelboss/featherlight/releases) or [the master including all the latest  bling](https://github.com/noelboss/featherlight/archive/master.zip).


# Installation

All styling is done using CSS so you'll want to include the Featherlight CSS in your head.

	<link href="//rawgithub.com/noelboss/featherlight/master/src/featherlight.min.css" type="text/css" rel="stylesheet" title="Featherlight Styles" />

Be aware that Featherlight uses very unspecific CSS selectors to help you overwrite every aspect. This means in turn, that if you're not following a modularized approach to write CSS (which you should! It's terrific!) and have many global and specific definitions (read ID's and such – which you shouldn't), these definitions can break the Featherlight styling.

Featherlight requires jQuery version 1.7.0 or higher. It's recommended to include the javascript at the bottom of the page before the closing `</body>` tag.

	<script src="//code.jquery.com/jquery-latest.js"></script>
	<script src="//rawgithub.com/noelboss/featherlight/master/src/featherlight.min.js" type="text/javascript" charset="utf-8"></script>


# Usage

By default, featherlight acts on all elements using the 'data-featherlight' attribute. An element with this attribute triggers the lightbox. The value of the attribute acts as selector for an element that's opened as lightbox.

	<a href="#" data-featherlight="#mylightbox">Open element in lightbox</a>
	<div id="mylightbox">This div will be opened in a lightbox</div>

Featherlight is smart. 'data-featherlight' can also contain a link to an image, an ajax-url or even DOM code:

	<span data-featherlight="myimage.png">Open image in lightbox</a>
	<span data-featherlight="myhtml.html .selector">Open ajax content in lightbox</a>
	<span data-featherlight="<p>Fancy DOM Lightbox!</p>">Open some DOM in lightbox</span>

it also works with links using href and the "image" and "ajax" keywords (this can also be manually set with the configuration options like `{image: 'photo.jpg}` or `{type: 'image'}`):

	<a href="myimage.png" data-featherlight="image">Open image in lightbox</a>
	<a href="myhtml.html .selector" data-featherlight="ajax">Open ajax content in lightbox</a>
	<a href="#" data-featherlight-ajax="myhtml.html .selector">Open ajax content in lightbox</a>
	<a href="#" data-featherlight="myhtml.html .selector" data-featherlight-type="ajax">Open ajax content in lightbox</a>

By default, Featherlight initializes all elements matching `defaults.selector` on document ready. If you want to prevent this, set `$.featherlight.defaults.autostart` to false before the DOM is ready.

## Bind Featherlight
You can bind the Featherlight events on any element using the following code:

	$('.myElement').featherlight($content, configuration);

It will then look for the `targetAttr` (by default "data-featherlight") on this element and use its value to find the content that will be opened as lightbox when you click on the element.

***configuration*** – Object: Object to configure certain aspects of the plugin. See [Configuration](#configuration).

***$content*** – jQuery Object or String: You can manually pass a jQuery object or a string (see [content filters](#content-filters)) to be opened in the ligthbox.

## Manual calling of Featherlight
In cases where you don't want an Element to act as Trigger you can call Featherlight manually. You can use this for example in an ajax callback to display the response data.

	$.featherlight($content, configuration);

***$content*** – jQuery Object or String: You can manually pass a jQuery object or a string (see [content filters](#content-filters)) to be opened in the ligthbox.

***configuration*** – Object: Object to configure certain aspects of the plugin. See [Configuration](#configuration).

# Configuration

Featherlight comes with a bunch of configuration-options which make it very flexible.
These options can be passed when calling `featherlight`.
Alternatively, they can specified as attribute on the elements triggering the lightbox;
for example, `<a data-featherlight-close-on-esc="false" ...>` has the same effect as
passing `{closeOnEsc: false}`.
You can also modify the `$.featherlight.defaults` directly which holds all the defaults:

	/* you can access and overwrite all defaults using $.featherlight.defaults */
	defaults: {
		namespace:    'featherlight',         /* Name of the events and css class prefix */
		targetAttr:   'data-featherlight',    /* Attribute of the triggered element that contains the selector to the lightbox content */
		variant:      null,                   /* Class that will be added to change look of the lightbox */
		resetCss:     false,                  /* Reset all css */
		background:   null,                   /* Custom DOM for the background, wrapper and the closebutton */
		openTrigger:  'click',                /* Event that triggers the lightbox */
		closeTrigger: 'click',                /* Event that triggers the closing of the lightbox */
		filter:       null,                   /* Selector to filter events. Think $(...).on('click', filter, eventHandler) */
		root:         'body',                 /* Where to append featherlights */
		openSpeed:    250,                    /* Duration of opening animation */
		closeSpeed:   250,                    /* Duration of closing animation */
		closeOnClick: 'background',           /* Close lightbox on click ('background', 'anywhere', or false) */
		closeOnEsc:   true,                   /* Close lightbox when pressing esc */
		closeIcon:    '&#10005;',             /* Close icon */
		otherClose:   null,                   /* Selector for alternate close buttons (e.g. "a.close") */
		beforeOpen:   $.noop,                 /* Called before open. can return false to prevent opening of lightbox. Gets event as parameter, this contains all data */
		beforeClose:  $.noop,                 /* Called before close. can return false to prevent opening of lightbox. Gets event as parameter, this contains all data */
		afterOpen:    $.noop,                 /* Called after open. Gets event as parameter, this contains all data */
		afterClose:   $.noop,                 /* Called after close. Gets event as parameter, this contains all data */
		type:         null,                   /* Specify content type. If unset, it will check for the targetAttrs value. */
		contentFilters: ['jquery', 'image', 'html', 'ajax', 'text'] /* List of content filters to use to determine the content */
		jquery/image/html/ajax/text: undefined     /* Specify content type and data */
	}

================================================

	namespace – String:  'featherlight'
All functions bound to elements are namespaced. This is also used to prefix all CSS classes for the background, the content-wrapper and the close button.


================================================

	targetAttr – String:  'data-featherlight'
Attribute on the triggering element pointing to the target element or content that will be opened in the lightbox.

================================================

	variant – String:  null
Pass your own CSS class to adjust the styling of the lightbox according to your need.


================================================

	resetCss – Boolean: false
Set this to true to remove all default css and start from designing scratch.

================================================


	openTrigger & closeTrigger – String:  'click'
Events that are used to open or close the lightbox. The close event is bound to the close button and to the lightbox background (if enabled).
Has no effect if $.featherlight is called directly.

================================================

	filter - String: null
A selector to filter events, when calling `featherlight` on a jQuery set, in a similar fashion to `$(...).on('click', filter, eventHandler)`.

Attributes both the selector and the filtered element are taken into account.

In the following example, the first link will make an ajax request while the second will display the text "second".

    <div data-featherlight data-featherlight-filter="a"
         data-featherlight-type="ajax">
      <a href="first">Hello</a>
      <a href="second" data-featherlight-type="text">World</a>
    </div>

Limitation: While auto bound elements added dynamically after onReady (e.g. via Ajax) will work fine, those with a `filter` are only supported supported if present before onReady.

================================================

	root - String: 'body'
This selector specified where the featherlight should be appended.

================================================

	openSpeed & closeSpeed – Integer or String:  250
Defines the speed for the opening and close animations. Values allowed are [jQuery animation durations](http://api.jquery.com/animate/#duration). Avoid the usage of 0 since this would cause a reversal of time and the end of the world! (Okay, kidding, just the closing callback would not be fired and unused ghost-DOM would linger around, but still.)


================================================

	closeOnClick – 'background', 'anywhere' or false
If set, the close event is also bound to the either the background only or anywhere

================================================

	closeOnEsc – Boolean: true
If true, the lightbox is closed when pressing the ESC key

================================================

	closeIcon – String: &#10005`;
Oh the naming...

================================================

  otherClose - String: null
While the close icon generated by featherlight will have class 'featherlight-close', you may specify alternate selector for other buttons having the same effect in your dialog, for example "a:contains('Cancel')".

================================================

	background – DOM String: null
You can provide the wrapping DOM. This is a bit tricky and just for the advanced users. It's recommended to study the plugin code. But you need to provide an element with a "{namespace}-close" class: the content of the lightbox will be added *after* this element.


================================================

	beforeOpen, beforeClose – Function: null
Called before the open or close method is executed. This function can return false to prevent open or
close method from execution. `this` is an object and contains the triggering DOM element (if existing) and the related Featherlight objects.

	// example
	beforeOpen: function(event){
		console.log(this); // this contains all related elements
		returm false; // prevent lightbox from opening
	}


================================================

	afterOpen, afterClose – Function: null
Called after the open or close method is executed – it is not called, if the `before-` or `open` function returns `false`! It receives the event object. `this` is an object and contains the triggering DOM element (if existing) and the related Featherlight objects.

	// example
	afterOpen: function(event){
		console.log(this); // this contains all related elements
		alert(this.$content.hasClass('true')); // alert class of content
	}

================================================

	type – String: null

The type object allows you to manually set what type the lightbox is. Set the value to 'image', 'ajax' or any of the content filters. Otherwise, the value from targetAttr will used be to determine the type of the lightbox. Example:

	$('.image-lightbox').featherlight({type: 'image'});


# Methods

`$.featherlight` is actually a constructor with `$.featherlight.methods` as a prototype and all the configuration options as attributes.

It's possible to use or change these methods, but the API isn't guaranteed to remain constant; enquire if you have particular needs.

	var current = $.featherlight.current()
	current.close();
	// do something else
	current.open(); // reopen it

Check the source code for more details.

# Globals

`$.featherlight` has the following globals:

	autoBind: '[data-featherlight]' /* Will automatically bind elements matching this selector. Clear or set before onReady */
	current: function() /* returns the currently opened featherlight, or undefined */
	close:   function() /* closes the currently opened featherlight (if any) */

# Content Filters

There are many ways to specify content to featherlight. Featherlight uses a set of heuristics to determine the type, for example data ending with `.gif` will be assumed to be an image. The following are equivalent:

	<a href="#" data-featherlight="photo.gif">See in a lightbox</a>

	<a href="photo.gif" data-featherlight>See in a lightbox</a>

	<a id="#example" href="#">See in a lightbox</a>
	<script>$('#example').featherlight('photo.gif')</script>

In case the heuristic wouldn't work, you can specify which contentFiter to use:

	<a href="photo_without_extension" data-featherlight="image">See in a lightbox</a>

	<a id="force_as_image" href="photo_without_extension">See in a lightbox</a>
	<script>
		$('#force_as_image').featherlight('image');
		// Equivalent:
		$('#force_as_image').featherlight({type: {image: true}});
	</script>

	<a id="force_as_image2" href="#">See in a lightbox</a>
	<script>$('#force_as_image2').featherlight('photo_without_extension', {type:{image: true}});</script>

You can add your own heuristics, for example:

	$.featherlight.contentFilters.feed = {
		regex: /^feed:/
		process: function(url) { /* deal with url */ return $('Loading...'); }
	}
	$.featherlight.defaults.contentFilters.unshift('feed');

This way the following would be possible:

	<a href="feed://some_url" data-featherlight>See the feed in a lightbox</a>

The content filter 'text' needs to be specified explicitly, it has no heuristic attached to it:

	<a href="Hello, world" data-featherlight="text">Example</a>

# Examples

## Use link-hashtags to open lightbox
	<a href="#targetElement" class="fl">Open</a>
	<div id="targetElement">Lightbox</div>

	$('a.fl').featherlight({
		targetAttr: 'href'
	});

## Open images with Featherlight
Us a link and point the data-featherlight attribute to the desired attribute which contains the link...

	<a href="myimage.jpg" data-featherlight="image">Open Image</a>

...or directly provide the link as the data-featherlight attribute:

	<span data-featherlight="myimage.jpg">Open Image</span>

## Open lightbox with ajax content
Use Featherlight with ajax using 'ajax' keyword or providing a url. It even supports selecting elements inside the response document.

	<a href="url.html .jQuery-Selector" data-featherlight="ajax">Open Ajax Content</a>

or you can provide the link directly as the featherlight-attribute:

	<a href="url.html .jQuery-Selector" data-featherlight="ajax">Open Ajax Content</a>


# IE8 background transparency
If you want the background in IE8 to be translucent, use data:image before the rgba background:


CHANGELOG.md
Featherlight – Changelog
===================================

Master
-----------------------------------
### Features
- New option filter to attach lazily like `$(sel).on('click', filter, ...)`. Attaching is done for the whole set, not on each individual element. Attributes of the whole set and the filtered
element are combined.
- New option otherClose to support extra closing buttons.
- New option root to support appending featherlight elsewhere than on the body.
- New content filter 'text' for plain text messages.
- Gallery has new callbacks `beforeImage` and `afterImage`.
- Gallery defaults are accessible with $.featherlightGallery.defaults
- Callbacks like afterOpen can be set using data-featherlight-after-open

### Breaking changes
- `$.featherlight` now always trigger a new lightbox. Use `$.fn.featherlight` for attaching events to elements.
- The order of the arguments of `$.fn.featherlight` has been reversed, for consistency and ease of use, but remain optional.
- The configuration options `open` and `close` have been removed.
- The configuration options `autostart`, `context` and `selector` have been merged to the global `$.featherlight.autoBind`. It's also lazily attached, so items added via ajax, for example, will be automatically attached too.
- The `$elm` attribute has been removed.
- The `config` attribute has been removed and merged with the `this` object.
- The `methods` global attribute has been removed. Methods have changed quite a bit. Access the prototype instead if needed.
- Content returned from filters isn't cloned anymore.

### Maintenance & Fixes
- Small bugs fixed
- Content filter 'jquery' now clones content *and* bound handlers.


0.4.4 - 2014-05-31
-----------------------------------


0.4.3 - 2014-05-31
-----------------------------------
### Maintenance & Fixes
- Call `afterClose` callback after the end of the fadeout

0.4.2 - 2014-05-29
-----------------------------------
### Maintenance & Fixes
- Avoid using global events since they are undocumented and apparently deprecated.

0.4.1 - 2014-04-17
-----------------------------------
### Maintenance & Fixes
- Fix dialog disappearing for large fixed width content
- Fig gallery navigation on IE

0.4.0 - 2014-04-07
-----------------------------------
### Features
- New option closeOnClick to support closing by clicking anywhere
– Any element inside the lightbox with the class "featherlight-close" will now trigger the closing event
- Support for nested lightboxes
- Gallery uses swipes on touch devices (if a supported library present)
- Gallery support and uses events to navigate through the images
- New functions: `$.featherlight.close()` and `current()`
- All configuration options have their equivalent html attribute.
- Improved and extensible content filters.

### Breaking changes
- `closeOnBg` no longer supported, use `closeOnClick`
- Gallery: class `.featherlight-prev` renamed to `.featherlight-previous`

The following changes could require changes to your custom CSS:

– Improves the way tall images are handled using a bottom border to maintain space (cut of instead of shrunken unproportionally)
– Improves the way the with of the lightbox is handled
– Changed paddings from % to pixels
– Changed closing icon class to featherlight-close-icon

### Maintenance & Fixes
- Improvements for mobile devices
- Many small bugs fixed
- Tests added
- $.fn.featherlight is now chainable



0.3.1 - 2014-03-01
-----------------------------------
### Maintenance & Fixes
- This.$content in after functions as well as open and close method now references actual elements inside the lightbox
- If open function has no content to open, afterOpen is prevented


0.3.0 - 2014-03-01
-----------------------------------
### Features
- Hooks! Added  before and after functions:
	´beforeClose´ ´beforeOpen´ ´afterOpen´ ´afterClose´
- Manually initialize all config.selector elements after config.autostart = false;


	$.featherlight();


### Maintenance & Fixes
- Take care, open and clouse defaults have been rewritten for Hooks!
- Preventing text selection
- Fixed issue #8
- Fixed issue #9


0.2.2 - 2014-02-10
-----------------------------------
### Features
- Added closeIcon option
- Added Gallery Configuration Options:


	gallery: {
		previous: '&#9664;',   /* Code that is used as previous icon */
		next: '&#9654;',       /* Code that is used as next icon */
		fadeIn: 100,           /* fadeIn speed when image is loaded */
		fadeOut: 300           /* fadeOut speed before image is loaded */
	}


0.2.1 - 2014-02-10
-----------------------------------
### Maintenance & Fixes
- Added image scale based on viewport height vh



0.2.0 - 2014-02-10
-----------------------------------
### Features
- Added Gallery Extension. Check out the [example page](http://noelboss.github.io/featherlight/gallery.html).


0.1.13 - 2014-02-09
-----------------------------------
### Maintenance & Fixes
- Fixed issue #7 where links inside featherlight don't work


0.1.12 - 2014-02-03
-----------------------------------
### Features
- Added configuration option 'openSpeed' and 'closeSpeed' to set animation duration

### Maintenance & Fixes
- Featherlight checks for jQuery and throws an error if not loaded
- Some jslint improvements
– Moved minified files to release/ folder
- Using grunt


0.1.11 - 2013-12-11
-----------------------------------
### Maintenance & Fixes
- Fixed iFrame detection
- Fixed close icon CSS


0.1.10 - 2013-12-10
-----------------------------------
### Features
- Added configuration object 'type' with two keys: type.image: false, type.ajax: false, to manually set type of lightbox

### Maintenance & Fixes
- Fixed behavior with missing data in targetAttr and handling of non found content


0.1.9 - 2013-12-05
-----------------------------------
### Features
- Added support for ESC key to close lightbox

### Maintenance & Fixes
- Renamed option clickBgClose to closeOnBg for consistency


0.1.8 - 2013-11-11
-----------------------------------

src/featherlight.js
/**
 * Featherlight - ultra slim jQuery lightbox
 * Version 0.4.4 - http://noelboss.github.io/featherlight/
 *
 * Copyright 2014, Noël Raoul Bossart (http://www.noelboss.com)
 * MIT Licensed.
**/
(function($) {
	"use strict";

	if('undefined' === typeof $) {
		if('console' in window){ window.console.info('Too much lightness, Featherlight needs jQuery.'); }
		return;
	}

	/* extend jQuery with standalone featherlight method  $.featherlight(elm, config); */
	function Featherlight($content, config) {
		if(this.constructor === Featherlight) {  /* called with new */
			this.id = Featherlight.id++;
			this.setup($content, config);
		} else {
			var fl = new Featherlight($content, config);
			fl.open();
			return fl;
		}
	}

		/* document wide esc handler, attached in setup method */
	var escapeHelper = function(event) {
			if (27 === event.keyCode && !event.isDefaultPrevented()) { // esc keycode
				var self = Featherlight.current();
				if(self && self.closeOnEsc) {
					self.$instance.find('.'+self.namespace+'-close:first').click();
					event.preventDefault();
				}
			}
		};

	Featherlight.prototype = {
		constructor: Featherlight,
		/*** defaults ***/
		/* extend featherlight with defaults and methods */
		namespace:    'featherlight',         /* Name of the events and css class prefix */
		targetAttr:   'data-featherlight',    /* Attribute of the triggered element that contains the selector to the lightbox content */
		variant:      null,                   /* Class that will be added to change look of the lightbox */
		resetCss:     false,                  /* Reset all css */
		background:   null,                   /* Custom DOM for the background, wrapper and the closebutton */
		openTrigger:  'click',                /* Event that triggers the lightbox */
		closeTrigger: 'click',                /* Event that triggers the closing of the lightbox */
		filter:       null,                   /* Selector to filter events. Think $(...).on('click', filter, eventHandler) */
		root:         'body',                 /* Where to append featherlights */
		openSpeed:    250,                    /* Duration of opening animation */
		closeSpeed:   250,                    /* Duration of closing animation */
		closeOnClick: 'background',           /* Close lightbox on click ('background', 'anywhere' or false) */
		closeOnEsc:   true,                   /* Close lightbox when pressing esc */
		closeIcon:    '&#10005;',             /* Close icon */
		otherClose:   null,                   /* Selector for alternate close buttons (e.g. "a.close") */
		beforeOpen:   $.noop,                 /* Called before open. can return false to prevent opening of lightbox. Gets event as parameter, this contains all data */
		beforeClose:  $.noop,                 /* Called before close. can return false to prevent opening of lightbox. Gets event as parameter, this contains all data */
		afterOpen:    $.noop,                 /* Called after open. Gets event as parameter, this contains all data */
		afterClose:   $.noop,                 /* Called after close. Gets event as parameter, this contains all data */
		type:         null,                   /* Specify type of lightbox. If unset, it will check for the targetAttrs value. */
		contentFilters: ['jquery', 'image', 'html', 'ajax', 'text'], /* List of content filters to use to determine the content */

		/*** methods ***/
		/* setup iterates over a single instance of featherlight and prepares the background and binds the events */
		setup: function(target, config){
			/* all arguments are optional */
			if (typeof target === 'object' && target instanceof $ === false && !config) {
				config = target;
				target = undefined;
			}

			var self = $.extend(this, config, {target: target}),
				css = !self.resetCss ? self.namespace : self.namespace+'-reset', /* by adding -reset to the classname, we reset all the default css */
				$background = $(self.background || '<div class="'+css+'"><div class="'+css+'-content"><span class="'+css+'-close-icon '+ self.namespace + '-close">'+self.closeIcon+'</span></div></div>'),
				closeButtonSelector = '.'+self.namespace+'-close' + (self.otherClose ? ',' + self.otherClose : '');

			self.$instance = $background.clone().addClass(self.variant); /* clone DOM for the background, wrapper and the close button */

			/* attach esc handler to document if configured */
			if(self.closeOnEsc && escapeHelper) {
				$(document).bind('keyup.'+this.constructor.prototype.namespace, escapeHelper);
				escapeHelper = null;
			}

			/* close when click on background/anywhere/null or closebox */
			self.$instance.on(self.closeTrigger+'.'+self.namespace, function(event) {
				var $target = $(event.target);
				if( ('background' === self.closeOnClick  && $target.is('.'+self.namespace))
					|| 'anywhere' === self.closeOnClick
					|| $target.is(closeButtonSelector) ){
					event.preventDefault();
					self.close();
				}
			});

			return this;
		},

		/* this method prepares the content and converts it into a jQuery object or a promise */
		getContent: function(){
			var self = this,
				filters = this.constructor.contentFilters,
				readTargetAttr = function(name){ return self.$currentTarget && self.$currentTarget.attr(name); },
				targetValue = readTargetAttr(self.targetAttr),
				data = self.target || targetValue || '';

			/* Find which filter applies */
			var filter = filters[self.type]; /* check explicit type like {type: 'image'} */

			/* check explicit type like data-featherlight="image" */
			if(!filter && data in filters) {
				filter = filters[data];
				data = self.target && targetValue;
			}
			data = data || readTargetAttr('href') || '';

			/* check explicity type & content like {image: 'photo.jpg'} */
			if(!filter) {
				for(var filterName in filters) {
					if(self[filterName]) {
						filter = filters[filterName];
						data = self[filterName];
					}
				}
			}

			/* otherwise it's implicit, run checks */
			if(!filter) {
				var target = data;
				data = null;
				$.each(self.contentFilters, function() {
					filter = filters[this];
					if(filter.test)  {
						data = filter.test(target);
					}
					if(!data && filter.regex && target.match && target.match(filter.regex)) {
						data = target;
					}
					return !data;
				});
				if(!data) {
					if('console' in window){ window.console.error('Featherlight: no content filter found ' + (target ? ' for "' + target + '"' : ' (no target specified)')); }
					return false;
				}
			}
			/* Process it */
			return filter.process.call(self, data);
		},

		/* sets the content of $instance to $content */
		setContent: function($content){
			var self = this;
			/* we need a special class for the iframe */
			if($content.is('iframe') || $('iframe', $content).length > 0){
				self.$instance.addClass(self.namespace+'-iframe');
			}
			self.$content = $content.addClass(self.namespace+'-inner');

			/* remove existing content */
			self.$instance.find('.'+self.namespace+'-inner').remove();
			self.$instance.find('.'+self.namespace+'-content').append(self.$content);

			return self;
		},

		/* opens the lightbox. "this" contains $instance with the lightbox, and with the config */
		open: function(event){
			var self = this;
			if(event && !event.isDefaultPrevented()
				|| self.beforeOpen(event) !== false) {

				if(event){
					event.preventDefault();
				}
				var $content = self.getContent();

				if($content){
					/* Add to opened registry */
					self.constructor._opened.add(self._openedCallback = function(response){
						if(self.$instance.closest('body').length > 0) {
							response.currentFeatherlight = self;
						}
					});

					/* Set content and show */
					$.when($content).done(function($content){
						self.setContent($content)
							.$instance.appendTo(self.root).fadeIn(self.openSpeed);
						self.afterOpen(event);
					});
					return self;
				}
			}
			return false;
		},

		/* closes the lightbox. "this" contains $instance with the lightbox, and with the config */
		close: function(event){
			var self = this;
			if(self.beforeClose(event) === false) {
				return false;
			}
			self.constructor._opened.remove(self._openedCallback);
			self.$instance.fadeOut(self.closeSpeed,function(){
				self.$instance.detach();
				self.afterClose(event);
			});
		}
	};

	$.extend(Featherlight, {
		id: 0,                                    /* Used to id single featherlight instances */
		autoBind:       '[data-featherlight]',    /* Will automatically bind elements matching this selector. Clear or set before onReady */
		defaults:       Featherlight.prototype,   /* You can access and override all defaults using $.featherlight.defaults, which is just a synonym for $.featherlight.prototype */
		/* Contains the logic to determine content */
		contentFilters: {
			jquery: {
				regex: /^[#.]\w/,         /* Anything that starts with a class name or identifiers */
				test: function(elem)    { return elem instanceof $ && elem; },
				process: function(elem) { return $(elem).clone(true); }
			},
			image: {
				regex: /\.(png|jpg|jpeg|gif|tiff|bmp)(\?\S*)?$/i,
				process: function(url)  {
					var self = this,
						deferred = $.Deferred(),
						img = new Image();
					img.onload  = function() { deferred.resolve(
						$('<img src="'+url+'" alt="" class="'+self.namespace+'-image" />')
					); };
					img.onerror = function() { deferred.reject(); };
					img.src = url;
					return deferred.promise();
				}
			},
			html: {
				regex: /^\s*<[\w!][^<]*>/, /* Anything that starts with some kind of valid tag */
				process: function(html) { return $(html); }
			},
			ajax: {
				regex: /./,            /* At this point, any content is assumed to be an URL */
				process: function(url)  {
					var self = this,
						deferred = $.Deferred();
					/* we are using load so one can specify a target with: url.html #targetelement */
					var $container = $('<div></div>').load(url, function(response, status){
						if ( status !== "error" ) {
							deferred.resolve($container.contents());
						}
						deferred.fail();
					});
					return deferred.promise();
				}
			},
			text: {
				process: function(text) { return $('<div>', {text: text}); }
			}
		},

		functionAttributes: ['beforeOpen', 'afterOpen', 'beforeClose', 'afterClose'],

		/*** class methods ***/
		/* read element's attributes starting with data-featherlight- */
		readElementConfig: function(element) {
			var Klass = this,
				config = {};
			if (element && element.attributes) {
					$.each(element.attributes, function(){
					var match = this.name.match(/^data-featherlight-(.*)/);
					if (match) {
						var val = this.value,
							name = $.camelCase(match[1]);
						if ($.inArray(name, Klass.functionAttributes) >= 0) {  /* jshint -W054 */
							val = new Function(val);                           /* jshint +W054 */
						} else {
							try { val = $.parseJSON(val); }
							catch(e) {}
						}
						config[name] = val;
					}
				});
			}
			return config;
		},

		attach: function($source, $content, config) {
			var Klass = this;
			if (typeof $content === 'object' && $content instanceof $ === false && !config) {
				config = $content;
				$content = undefined;
			}
			/* make a copy */
			config = $.extend({}, config);

			/* Only for openTrigger and namespace... */
			var tempConfig = $.extend({}, Klass.defaults, Klass.readElementConfig($source[0]), config);

			$source.on(tempConfig.openTrigger+'.'+tempConfig.namespace, tempConfig.filter, function(event) {
				/* ... since we might as well compute the config on the actual target */
				var elemConfig = $.extend({$currentTarget: $(this)}, Klass.readElementConfig($source[0]), Klass.readElementConfig(this), config);
				new $.featherlight($content, elemConfig).open(event);
			});
		},

		current: function() {
			var response = {};
			this._opened.fire(response);
			return response.currentFeatherlight;
		},

		close: function() {
			var cur = this.current();
			if(cur) { cur.close(); }
		},

		_opened: $.Callbacks()
	});

	$.featherlight = Featherlight;

	/* bind jQuery elements to trigger featherlight */
	$.fn.featherlight = function($content, config) {
		Featherlight.attach(this, $content, config);
		return this;
	};

	/* bind featherlight on ready if config autoBind is set */
	$(document).ready(function(){
		if(Featherlight.autoBind){
			/* First, bind click on document, so it will work for items added dynamically */
			$(document).featherlight({filter: Featherlight.autoBind});
			/* Auto bound elements with attr-featherlight-filter won't work
			   (since we already used it to bind on document), so bind these
			   directly. We can't easily support dynamically added element with filters */
			$(Featherlight.autoBind).filter('[data-featherlight-filter]').each(function(){
				$(this).featherlight();
			});
		}
	});
}(jQuery));
